### ISSUE TITLE
Transform zeros to `NULLS` for `household_pseudo_id` and `household_size`

### ISSUE THREAD

**Issue by iaindillingham** [opensafely-core/ehrql#1771]
[In Cohort Extractor][1], `patients.household_as_of(returning="pseudo_id")` and `patients.household_as_of(returning="household_size")` return zero if no household information is available.

In ehrQL, `household_memberships_2020.household_pseudo_id` and `household_memberships_2020.household_size` also return zero if no household information is available. However, they should return `NULL`, such that a user can benefit from `is_null`, `is_not_null`, etc.

Unfortunately, I don't think there is a workaround, because

```py
nullable_household_size = case(
    when(household_memberships_2020.household_size == 0).then(None),
    default=household_memberships_2020.household_size,
)
```

raises a `TypeValidationError`.

Thanks for raising this in Slack,[^1] @wjchulme and @emprestige.

---

Minimal `dataset_definition.py`:

```py
from ehrql import case, create_dataset, when
from ehrql.tables.beta.core import patients
from ehrql.tables.beta.tpp import household_memberships_2020

nullable_household_size = case(
    when(household_memberships_2020.household_size == 0).then(None),
    default=household_memberships_2020.household_size,
)

dataset = create_dataset()
dataset.define_population(patients.exists_for_patient())
dataset.configure_dummy_data(population_size=10)

dataset.household_size = household_memberships_2020.household_size

dataset.nullable_household_size = nullable_household_size
```

Output from running `opensafely exec ehrql:v0 generate-dataset dataset_definition.py`:

```
2023-11-23 17:57:25 [info     ] Compiling dataset definition from dataset_definition.py
Failed to import 'dataset_definition.py':

Traceback (most recent call last):
  File "/workspace/dataset_definition.py", line 5, in <module>
    nullable_household_size = case(
                              ^^^^^
ehrql.query_model.nodes.TypeValidationError: Case.default requires 'ehrql.query_model.nodes.Series[NoneType] | None' but got 'ehrql.query_model.nodes.Series[int]'
```

[1]: https://docs.opensafely.org/study-def-variables/#cohortextractor.patients.household_as_of

[^1]: https://bennettoxford.slack.com/archives/C04DVD1UQC9/p1700753981157869

**Comment by evansd** [comment]
> Unfortunately, I don't think there is a workaround ...

It is actually possible to work around this, though I agree that we should handle this automatically in the backend. The problem you're hitting is that you can't introduce explicit NULLs into the query model. But you can use NULL as a default so you just need to reverse the condition:
```python
nullable_household_size = case(
    when(household_memberships_2020.household_size != 0).then(household_memberships_2020.household_size),
    default=None,
)
```

Or alternatively:
```python
nullable_household_size = (
    when(household_memberships_2020.household_size != 0)
    .then(household_memberships_2020.household_size)
    .otherwise(None)
)
```

**Comment by inglesp** [comment]
Merged into v1 branch in #1811.
